File: app/src/main/java/com/zaclimon/acetv/ui/settings/EpgForceSyncGuidedFragment.java
package com.zaclimon.acetv.ui.settings;

import android.content.ComponentName;
import android.content.ContentResolver;
import android.media.tv.TvContract;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v17.leanback.app.GuidedStepSupportFragment;
import android.support.v17.leanback.widget.GuidanceStylist;
import android.support.v17.leanback.widget.GuidedAction;

import com.google.android.media.tv.companionlibrary.EpgSyncJobService;
import com.google.android.media.tv.companionlibrary.model.Channel;
import com.google.android.media.tv.companionlibrary.utils.TvContractUtils;
import com.zaclimon.acetv.R;
import com.zaclimon.acetv.service.AceJobService;
import com.zaclimon.acetv.util.Constants;

import java.util.List;
import java.util.concurrent.TimeUnit;

/**
 * Setting option that forces an EPG sync for the next hour.
 *
 * @author zaclimon
 * Creation date: 22/06/17
 */

public class EpgForceSyncGuidedFragment extends GuidedStepSupportFragment {

    @Override
    public GuidanceStylist.Guidance onCreateGuidance(Bundle savedInstanceState) {
        String title = getString(R.string.force_epg_sync_title);
        String description = getString(R.string.force_epg_sync_description);

        return (new GuidanceStylist.Guidance(title, description, null, null));
    }

    @Override
    public void onCreateActions(List<GuidedAction> actions, Bundle savedInstanceState) {
        GuidedAction.Builder yesAction = new GuidedAction.Builder(getActivity());
        GuidedAction.Builder noAction = new GuidedAction.Builder(getActivity());
        yesAction.clickAction(GuidedAction.ACTION_ID_YES);
        noAction.clickAction(GuidedAction.ACTION_ID_NO);
        actions.add(yesAction.build());
        actions.add(noAction.build());
    }

    @Override
    public void onGuidedActionClicked(GuidedAction guidedAction) {
        if (guidedAction.getId() == GuidedAction.ACTION_ID_YES) {
            new AsyncResyncPrograms().execute();
            add(getFragmentManager(), new EpgSyncLoadingGuidedFragment());
        } else {
            getActivity().finish();
        }
    }

    /**
     * Private class that removes all programs and then syncs them back up
     *
     * @author zaclimon
     * Creation date: 27/08/17
     */
    private class AsyncResyncPrograms extends AsyncTask<Void, Void, Void> {

        @Override
        public Void doInBackground(Void... params) {

            if (isAdded()) {
                ContentResolver contentResolver = getActivity().getContentResolver();
                List<Channel> channels = TvContractUtils.getChannels(contentResolver);

                for (Channel channel : channels) {
                    Uri channelProgramsUri = TvContract.buildProgramsUriForChannel(channel.getId());
                    contentResolver.delete(channelProgramsUri, null, null);
                }
            }
            return (null);
        }

        @Override
        public void onPostExecute(Void result) {
            String inputId = TvContract.buildInputId(Constants.TV_INPUT_SERVICE_COMPONENT);
            EpgSyncJobService.requestImmediateSync(getActivity(), inputId, TimeUnit.HOURS.toMillis(48), new ComponentName(getActivity(), AceJobService.class));
        }
    }

}

File: app/src/main/java/com/zaclimon/tsutaeru/ui/settings/EpgForceSyncGuidedFragment.kt
package com.zaclimon.tsutaeru.ui.settings

import android.content.ComponentName
import android.content.Context
import android.media.tv.TvContract
import android.os.AsyncTask
import android.os.Bundle
import android.support.v17.leanback.app.GuidedStepSupportFragment
import android.support.v17.leanback.widget.GuidanceStylist
import android.support.v17.leanback.widget.GuidedAction
import com.google.android.media.tv.companionlibrary.model.ModelUtils
import com.google.android.media.tv.companionlibrary.sync.EpgSyncJobService
import com.google.android.media.tv.companionlibrary.utils.TvContractUtils
import com.zaclimon.tsutaeru.R
import com.zaclimon.tsutaeru.service.TsutaeruJobService
import com.zaclimon.tsutaeru.util.Constants
import java.lang.ref.WeakReference

/**
 * Setting option that forces an EPG sync for the next hour.
 *
 * @author zaclimon
 */
class EpgForceSyncGuidedFragment : GuidedStepSupportFragment() {

    override fun onCreateGuidance(savedInstanceState: Bundle?): GuidanceStylist.Guidance {
        val title = getString(R.string.force_epg_sync_title)
        val description = getString(R.string.force_epg_sync_description)

        return GuidanceStylist.Guidance(title, description, null, null)
    }

    override fun onCreateActions(actions: MutableList<GuidedAction>, savedInstanceState: Bundle?) {
        val yesAction = GuidedAction.Builder(context).apply {
            clickAction(GuidedAction.ACTION_ID_YES)
        }.build()

        val noAction = GuidedAction.Builder(context).apply {
            clickAction(GuidedAction.ACTION_ID_NO)
        }.build()

        actions.add(yesAction)
        actions.add(noAction)
    }

    override fun onGuidedActionClicked(action: GuidedAction?) {
        val id = action?.id

        if (id == GuidedAction.ACTION_ID_YES) {
            context?.let {
                AsyncResyncPrograms(it).execute()
                add(fragmentManager, EpgSyncLoadingGuidedFragment())
            }
        } else {
            activity?.finish()
        }
    }

    private class AsyncResyncPrograms(context: Context) : AsyncTask<Void, Void, Void?>() {

        private val asyncReference = WeakReference<Context>(context)

        override fun doInBackground(vararg p0: Void?): Void? {
            val context = asyncReference.get()
            val contentResolver = context?.contentResolver
            val channels = ModelUtils.getChannels(contentResolver)

            for (channel in channels) {
                val programsUri = TvContract.buildProgramsUriForChannel(channel.id)
                contentResolver?.delete(programsUri, null, null)
            }
            return null
        }

        override fun onPostExecute(result: Void?) {
            val context = asyncReference.get()
            val inputId = TvContract.buildInputId(Constants.TV_INPUT_SERVICE_COMPONENT)
            EpgSyncJobService.requestImmediateSync(context, inputId, ComponentName(context, TsutaeruJobService::class.java))
        }
    }

}